### PR TITLE
refactor: Migrate ccb to new make learner structure

### PR DESCRIPTION
**Pull request by jackgerrits**

### CODE REVIEW

File: vowpalwabbit/conditional_contextual_bandit.cc [hunk 1]
```diff
@@ -698,12 +698,15 @@ base_learner* ccb_explore_adf_setup(options_i& options, vw& all)
...
-      prediction_type_t::decision_probs, all.get_setupfn_name(ccb_explore_adf_setup), true);
-
-  l.set_finish_example(finish_multiline_example);
-  l.set_save_load(save_load);
-  return make_base(l);
+  auto* l = VW::LEARNER::make_reduction_learner(std::move(data), base, learn_or_predict<true>, learn_or_predict<false>,
+      all.get_setupfn_name(ccb_explore_adf_setup))
+                .set_learn_returns_prediction(true)
+                .set_prediction_type(prediction_type_t::decision_probs)
+                .set_label_type(label_type_t::ccb)
```
Comment (lalo) [inline vowpalwabbit/conditional_contextual_bandit.cc:705]: so label type is a new addition, i'm not seeing on the left side

File: vowpalwabbit/conditional_contextual_bandit.cc [hunk 2]
```diff
@@ -698,12 +698,15 @@ base_learner* ccb_explore_adf_setup(options_i& options, vw& all)
   data->id_namespace_str.append("_id");
   data->id_namespace_hash = VW::hash_space(all, data->id_namespace_str);
 
-  learner<ccb, multi_ex>& l = init_learner(data, base, learn_or_predict<true>, learn_or_predict<false>, 1,
```
Comment (lalo) [inline vowpalwabbit/conditional_contextual_bandit.cc:701]: where is the 1 specified now?

File: vowpalwabbit/conditional_contextual_bandit.cc (hunk 1 above)
Comment (jackgerrits) [reply vowpalwabbit/conditional_contextual_bandit.cc:705]: Correct, this is something new that learners should specify

File: vowpalwabbit/conditional_contextual_bandit.cc (hunk 2 above)
Comment (jackgerrits) [reply vowpalwabbit/conditional_contextual_bandit.cc:701]: 1 is the default and so it doesnt need to be added in that case

File: vowpalwabbit/conditional_contextual_bandit.cc (hunk 1 above)
Comment (olgavrou) [reply vowpalwabbit/conditional_contextual_bandit.cc:705]: this information is also specified by the label parser, maybe we should get it from there? (i.e. here do: `all.example_parser->lbl_parser.label_type`)

Also could we use that info from the label parser? where will the learner use it?

Comment (jackgerrits) [reply vowpalwabbit/conditional_contextual_bandit.cc:705]: The label parser can only be used for the label type of the topmost reduction and not each one in the stack

The learner is going to use this to enforce type safety between reductions in the long term

Comment (lalo) [reply vowpalwabbit/conditional_contextual_bandit.cc:705]: I think we should add ac omment on set_label_type clarifying Olga's point and mnentioning that its not used for now and its for long term design.

Comment (olgavrou) [reply vowpalwabbit/conditional_contextual_bandit.cc:705]: maybe in the set function?

**Review by olgavrou — APPROVED** [review]

File: vowpalwabbit/conditional_contextual_bandit.cc (hunk 1 above)
Comment (jackgerrits) [reply vowpalwabbit/conditional_contextual_bandit.cc:705]: There already was a comment in the set function. I added a bit to it stating why it is helpful

**Pull request merged**